### PR TITLE
ARC-404: Adding CICD to micros branch.

### PR DESCRIPTION
**Pull request by atrigueiro**

### CODE REVIEW

File: src/github/index.ts [hunk 1]
```diff
@@ -49,6 +51,10 @@ export default (robot) => {
 		middleware(pullRequest)
 	);
 
+	robot.on("workflow_run", middleware(workflow));
+
+	robot.on("deployment_status", middleware(deployment));
```
Comment (mboudreau) [inline src/github/index.ts:56]: As per your changes above, shouldn't there be a `deployment` event as well?

File: src/github/middleware.ts [hunk 2]
```diff
@@ -82,16 +82,12 @@ export default (
...
 		) {
-			context.log(
-				{
-					noop: "bot",
-					botId: context.payload.sender.id
-				},
-				"Halting further execution since the sender is a bot and action is not a state change"
-			);
-			return;
+			if (!isStateChangeOrDeploymentAction(context.name)) {
```
Comment (mboudreau) [inline src/github/middleware.ts:87]: shouldn't these 2 conditions be merged into a single if statement? or is there something missing after the second if?

File: src/jira/client/axios.ts [hunk 3]
```diff
@@ -114,7 +114,7 @@ function getSuccessMiddleware(logger) {
 				},
 				`Jira request: ${response.config.method.toUpperCase()} ${
 					response.config.originalUrl
-				} - ${response.status} ${response.statusText}`
+				} - ${response.status} ${response.statusText}\\n${JSON.stringify(response.data)}\`);`
```
Comment (mboudreau) [inline src/jira/client/axios.ts:117]: the JSON.stringify bit doesn't look right.  First, you have a `\\n` which will escape and show literally `\n` in the string.  Then you have an errant escaped tilde and a `);` in the string.

File: src/transforms/deployment.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+			return "successful";
+		case "error":
+		case "failure":
+			return "failed";
+		default:
+			return "unknown";
+	}
+}
+
+function mapEnvironment(environment: string) {
```
Comment (mboudreau) [inline src/transforms/deployment.ts:26]: same here.

File: src/transforms/deployment.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
+import { Context } from "probot/lib/context";
+import issueKeyParser from "jira-issue-key-parser";
+
+// We need to map the state of a GitHub deployment back to a valid deployment state in Jira.
+// https://docs.github.com/en/rest/reference/repos#list-deployments
+// Deployment state - GitHub: Can be one of error, failure, pending, in_progress, queued, or success
+// https://developer.atlassian.com/cloud/jira/software/rest/api-group-builds/#api-deployments-0-1-bulk-post
+// Deployment state - Jira: Can be one of unknown, pending, in_progress, cancelled, failed, rolled_back, successful
+function mapState(state: string) {
```
Comment (mboudreau) [inline src/transforms/deployment.ts:9]: I'd probably set the return type here.

File: src/transforms/deployment.ts [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+	const jiraEnv = Object.keys(environmentMapping).find(key => isEnvironment(environmentMapping[key]));
+
+	if (!jiraEnv) {
+		return "unmapped";
+	}
+
+	return jiraEnv;
+}
+
+export default async (context: Context) => {
```
Comment (mboudreau) [inline src/transforms/deployment.ts:50]: Would be great to get the return type here.

File: src/github/deployment.ts [hunk 7]
```diff
@@ -0,0 +1,13 @@
+import transformDeployment from "../transforms/deployment";
+import { Context } from "probot/lib/context";
+
+export default async (context: Context, jiraClient) => {
+	const { data: jiraPayload } = await transformDeployment(context);
```
Comment (mboudreau) [inline src/github/deployment.ts:5]: Is there a reason that we call a function that returns an object with a key of `data` that we then completely ignore that data key?  Why doesn't the function just return the jiraPayload object?  No need to faff about.

File: src/github/workflow.ts [hunk 8]
```diff
@@ -0,0 +1,13 @@
+import transformWorkflow from "../transforms/workflow";
+import { Context } from "probot/lib/context";
+
+export default async (context: Context, jiraClient) => {
+	const { data: jiraPayload } = await transformWorkflow(context);
```
Comment (mboudreau) [inline src/github/workflow.ts:5]: Same problem here, we completely discard the `data` key, hence we don't need to have it in the function either.

File: src/transforms/workflow.ts [hunk 9]
```diff
@@ -0,0 +1,75 @@
+import issueKeyParser from "jira-issue-key-parser";
+import { Context } from "probot/lib/context";
+
+// We need to map the status and conclusion of a GitHub workflow back to a valid build state in Jira.
+// https://docs.github.com/en/rest/reference/actions#list-workflow-runs-for-a-repository
+// Workflow status - GitHub: Can be one of queued, in_progress, or completed.
+// Workflow conclusion - GitHub: Can be one of action_required, cancelled, failure, neutral, success, skipped, stale, or timed_out
+// https://developer.atlassian.com/cloud/jira/software/rest/api-group-builds/#api-builds-0-1-bulk-post
+// Build state - Jira: Can be one of pending, in_progress, successful, failed, cancelled, unknown
+function mapStatus(status: string, conclusion: string) {
```
Comment (mboudreau) [inline src/transforms/workflow.ts:10]: return type and conclusion is optional

File: src/transforms/workflow.ts [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+		case "completed.stale":
+			return "cancelled";
+		case "completed.action_required":
+			return "pending";
+		default:
+			return "unknown";
+	}
+}
+
+function mapPullRequests(pull_requests) {
```
Comment (mboudreau) [inline src/transforms/workflow.ts:34]: parameter and return type would be nice.

**Review by mboudreau — CHANGES_REQUESTED** [review]
Looking good, just need a few typings here and there :)

File: src/github/branch.ts [hunk 11]
```diff
@@ -9,7 +9,7 @@ export const createBranch = async (context: Context, jiraClient) => {
 	if (!jiraPayload) {
 		context.log(
 			{ noop: "no_jira_payload_create_branch" },
-			"Halting futher execution for createBranch since jiraPayload is empty"
+			"Halting further execution for createBranch since jiraPayload is empty"
```
Comment (KAbakumovAtl) [inline src/github/branch.ts:12]: nice!

File: src/jira/client/index.ts [hunk 12]
```diff
@@ -213,7 +213,46 @@ async function getJiraClient(
...
+					const subscription = await Subscription.getSingleInstallation(jiraHost, gitHubInstallationId);
+					await subscription.update({ syncWarning: "Exceeded issue key reference limit. Some issues may not be linked." });
+				}
+				const payload = {
+					deployments: data.deployments,
+					properties: {
+						gitHubInstallationId,
+					},
+				};
+				logger.info(`Sending deployments payload to jira. Payload: ${payload}`);
```
Comment (KAbakumovAtl) [inline src/jira/client/index.ts:252]: I don't think we should log payloads hence they contain UGC. If we need it for testing we should change to `debug`

- [ ] remove payloads logging in prod

Comment (KAbakumovAtl) [reply src/jira/client/index.ts:252]: Still good to have a logline which indicates that the request was made. So we just need to remove the ${payload} part

File: src/github/index.ts (hunk 1 above)
Comment (atrigueiro) [reply src/github/index.ts:56]: Nope, we only listen to `deployment_status`. I'll comment the `deployment` event from https://github.com/atlassian/github-for-jira/pull/573/files#diff-4f81d94913f59db9100bd467a208d1a30a2c9d548aa71d6a54e4fd710c3a3018R21

File: src/jira/client/axios.ts (hunk 3 above)
Comment (atrigueiro) [reply src/jira/client/axios.ts:117]: Copying/pasting fail 🤦‍♀️

File: src/github/deployment.ts (hunk 7 above)
Comment (atrigueiro) [reply src/github/deployment.ts:5]: No need, returning only jiraPayload now.

File: src/transforms/commit.ts [hunk 13]
```diff
@@ -29,33 +30,6 @@ function mapCommit(githubCommit, author): Commit {
 	};
 }
 
-interface Commit {
```
Comment (atrigueiro) [inline src/transforms/commit.ts:32]: Moved to new file `src/interfaces/jira.ts`

File: src/transforms/deployment.ts [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+			return "in_progress";
+		case "success":
+			return "successful";
+		case "error":
+		case "failure":
+			return "failed";
+		default:
+			return "unknown";
+	}
+}
```
Comment (atrigueiro) [inline src/transforms/deployment.ts:24]: Don't know why the tabbing looks messed here but it looks good on my IDE
Screen Shot 2021-08-09 at 1 20 21 pm

File: src/github/deployment.ts [hunk 15]
```diff
@@ -0,0 +1,13 @@
+import transformDeployment from "../transforms/deployment";
+import { Context } from "probot/lib/context";
+
+export default async (context: Context, jiraClient) => {
```
Comment (mboudreau) [inline src/github/deployment.ts:4]: Should have a return type of `:Promise<void>`

File: src/github/workflow.ts [hunk 16]
```diff
@@ -0,0 +1,13 @@
+import transformWorkflow from "../transforms/workflow";
+import { Context } from "probot/lib/context";
+
+export default async (context: Context, jiraClient) => {
```
Comment (mboudreau) [inline src/github/workflow.ts:4]: return `Promise<void>`

File: src/interfaces/jira.ts [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+		id: string;
+		repositoryUri: string;
+	};
+	ref: {
+		name: string;
+		uri: string;
+	}
+}
+
+export interface Build {
```
Comment (mboudreau) [inline src/interfaces/jira.ts:12]: Probably be good to prefix all interfaces with `Jira` to not cause any confusion.

File: src/transforms/deployment.ts (hunk 14 above)
Comment (mboudreau) [reply src/transforms/deployment.ts:25]: it's fine :)
It's because it's using tabs and for some reason github shows tabs as 8 spaces, which is massive...

**Review by mboudreau — APPROVED** [review]
:crossed_fingers:

**Pull request merged**